File: 53MaxSubarray.cs
public class Solution {
        public int MaxSubArray(int[] nums) 
        {
            if (nums.Length == 0) return 0;
            var curr = nums[0];
            var max = curr;
            for (int i = 0; i < nums.Length; i++) 
            {
                // if current is losing value, replace
                if (nums[i] > curr + nums[i])
                    curr = nums[i];
                else 
                    curr = curr + nums[i];
                
                max = curr > max ? curr : max;
            }
            return max;
        }

}

File: 53maxsubarray.py
class Solution:
    def maxSubArray(self, nums: []) -> int:
        maxSum = currSum = nums[0]
        for i in range(1, len(nums)):
            # always reset when our running sum dips too low
            currSum = max(nums[i], currSum + nums[i])
            # compare to global max
            maxSum = max(currSum, maxSum)
            
        return maxSum
            
